src/main/java/com/minecolonies/coremod/colony/buildings/inventory/HashSetIndex.java

```diff
@@ -0,0 +1,124 @@
+package com.minecolonies.coremod.colony.buildings.inventory;
+
+import com.minecolonies.api.util.Tuple;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+
+/**
+ * Generic index class which holds an index of type Key and a Set of Values.
+ */
+public class HashSetIndex<Key, Value>
```

Comment:
This is complete overkill and not needed there are several classes that can do this job in the JSL or even in Guava.
As such this class can be removed and its logic replaced by one on of them.

A good replacement would be a MultiMap.

Reply:
Also this follows no JSL interface for generic collections, so that is needed as well!